### ISSUE TITLE
Putting Craft 3 in developer mode throws Yii errors relating to Redis

### ISSUE THREAD

**Issue by whereisk** [wyveo/craftcms-docker#12]
As the title sugggests, Yii throws errors when putting Craft3 in developer mode (when accessing the control panel).
yii\base\ErrorException: session_start(): Failed to initialize storage module: redis
I've put the stack trace here: https://pastebin.com/qVaz6g77

**Comment by colinwilson** [comment]
Having trouble recreating this. Can you provide a little more info on how you're deploying?

**Comment by colinwilson** [comment]
Closing due to lack of response.

**Comment by whereisk** [comment]
Apologies, I just saw this. I'm not sure what extra information you are
requiring? Do you mean the environment of deployment?


On Thu, 10 May 2018, at 6:32 AM, Colin Wilson wrote:
> Having trouble recreating this. Can you provide a little more info on
> how you're deploying?> — You are receiving this because you authored the thread. Reply to
> this email directly, view it on GitHub[1], or mute the thread[2].> 


Links:

  1. https://github.com/wyveo/craftcms-docker/issues/12#issuecomment-387866562

**Comment by colinwilson** [comment]
When I tried to reproduce this I deployed using the `docker-compose.yml` file, set devMode to `true` in  `config/general.php` and could access the control panel without issue/error.

Just wondering if you're deploying using the same method?

**Comment by whereisk** [comment]
Ah, I see, no what I did was ssh'd into the docker instance and changed
the config file from there.Then it only fails when I log-in to the admin interface. The front end
is working fine.
Let me know if you can access the admin interface as expected with the
method you described.


On Tue, 15 May 2018, at 1:40 PM, Colin Wilson wrote:
> When I tried to reproduce this I deployed using the docker-compose.yml
> file, set devMode to true in config/general.php and could access the
> control panel without issue/error.> Just wondering if you're deploying using the same method?


> — You are receiving this because you authored the thread. Reply to
> this email directly, view it on GitHub[1], or mute the thread[2].> 


Links:

  1. https://github.com/wyveo/craftcms-docker/issues/12#issuecomment-389033079

**Comment by whereisk** [comment]
As a follow up, I destroyed and re-created the docker image after changing the config file but it didn't flow through into the install. Ssh'ing into the docker instance to /usr/share/nginx/config/general.php it still shows dev-mode false, and it fails to load the admin interface if changed to true. 

Let me know if that doesn't make sense.

**Comment by colinwilson** [comment]
> Let me know if you can access the admin interface as expected with the method you described.

Yep.

Try my (docker-compose) method. From your stack trace it looks like you may have an issue connecting to the redis container.

**Comment by whereisk** [comment]
So I did that, but the config file inside the docker instance did not
take the changes from the external config.php (you can tell when Craft
is in developer mode by some stripy lines when in the Control Panel).
Are you sure your instance is in developer mode?
On Tue, 15 May 2018, at 11:01 PM, Colin Wilson wrote:
>> Let me know if you can access the admin interface as expected with
>> the method you described.> Yep.


> Try my (docker-compose) method. From you stack trace it looks like you
> may have an issue connecting to the redis container instance.> — You are receiving this because you authored the thread. Reply to
> this email directly, view it on GitHub[1], or mute the thread[2].> 


Links:

  1. https://github.com/wyveo/craftcms-docker/issues/12#issuecomment-389157892

**Comment by colinwilson** [comment]
> Are you sure your instance is in developer mode?

Positive:

![2018-05-16 05_56_26-rancher](https://user-images.githubusercontent.com/12916656/40097505-68910a2c-58ce-11e8-8816-aacb51ecd518.png)
![2018-05-16 05_58_43-dashboard - test](https://user-images.githubusercontent.com/12916656/40097512-6bb66e68-58ce-11e8-8610-629fd3b77bef.png)


> did not take the changes from the external config.php

You mean `config/general.php`?

**Comment by whereisk** [comment]
Yeah, fair enough. But it definitely is not doing it here.
I deleted all the images and tried again.

Perhaps something with this docker version? What host system are you
using docker on?I'm running on a VM of Ubuntu18.04 - which doesn't have an official
version for docker yet, so maybe it has to do with that.

On Wed, 16 May 2018, at 3:07 PM, Colin Wilson wrote:
>> Are you sure your instance is in developer mode?


> Positive:


> 2018-05-16 05_56_26-rancher 2018-05-16 05_58_43-dashboard - test


>> did not take the changes from the external config.php


> You mean config/general.php?


> — You are receiving this because you authored the thread. Reply to
> this email directly, view it on GitHub[1], or mute the thread[2].> 


Links:

  1. https://github.com/wyveo/craftcms-docker/issues/12#issuecomment-389395706

**Comment by colinwilson** [comment]
what's the result of `# docker -v`? I'll try to recreate your environment for testing.

**Comment by whereisk** [comment]
Thank you. Here is the outcome of docker -v:
Docker version 17.12.1-ce, build 7390fc6



On Sat, 19 May 2018, at 12:32 AM, Colin Wilson wrote:
> what's the result of # docker -v? I'll try to recreate your
> environment for testing.> — You are receiving this because you authored the thread. Reply to
> this email directly, view it on GitHub[1], or mute the thread[2].> 


Links:

  1. https://github.com/wyveo/craftcms-docker/issues/12#issuecomment-390226385

**Comment by mitchhh22** [comment]
Getting the same error when setting devMode true. 
I'm running the most recent commit (9375ed1) and up to date docker on OSX.

**Comment by bschulz87** [comment]
Also getting error: session_start(): Failed to initialize storage module: redis (path: ) after setting devMode => true in config/general.php

Setting to false again makes the site work as expected.

docker -v
Docker version 18.03.1-ce, build 9ee9f40

docker-compose -v
docker-compose version 1.21.2, build a133471

**Comment by Yugloocamai** [comment]
It looks like the Yii Session class is being used directly in Craft3 by way of Craft's Application class extending the Yii Application class.  This means that Yii's Session->getUseCustomStorage method isn't being overridden.  This method returns false by default, which will prevent registerSessionHandler from doing anything.  

**yii2/web/Session.php:**
`/**
     * Returns a value indicating whether to use custom session storage.
     * This method should be overridden to return true by child classes that implement custom session storage.
     * To implement custom session storage, override these methods: [[openSession()]], [[closeSession()]],
     * [[readSession()]], [[writeSession()]], [[destroySession()]] and [[gcSession()]].
     * @return bool whether to use custom storage.
     */`

    public function getUseCustomStorage()
    {
        return false;
    }

Since Craft3 is using that class directly and not overriding any functions, it seems that any attempt to use custom session configurations will fail.  This is probably why Redis config fails in devmode.  In non-devmode the @ sign is prepended to the session_start call which suppresses warnings.  A quick workaround would be to simply change return false; to return true;  This seems like a Craft3 bug to me though.
<br><br>
TL;DR
Change **return false;** to **return true;** on line 121 of **yii2/web/Session.php**
